subprojects/workers/src/main/java/org/gradle/workers/internal/IsolatedClassloaderWorker.java

```diff
@@ -0,0 +1,184 @@
+/*
+ * Copyright 2019 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.workers.internal;
+
+import org.gradle.api.internal.classloading.GroovySystemLoader;
+import org.gradle.api.internal.classloading.GroovySystemLoaderFactory;
+import org.gradle.internal.UncheckedException;
+import org.gradle.internal.classloader.ClassLoaderSpec;
+import org.gradle.internal.classloader.FilteringClassLoader;
+import org.gradle.internal.classloader.VisitableURLClassLoader;
+import org.gradle.internal.concurrent.CompositeStoppable;
+import org.gradle.internal.io.ClassLoaderObjectInputStream;
+import org.gradle.internal.serialize.ExceptionReplacingObjectInputStream;
+import org.gradle.internal.serialize.ExceptionReplacingObjectOutputStream;
+import org.gradle.internal.service.ServiceRegistry;
+import org.gradle.util.GUtil;
+
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.ObjectInputStream;
+import java.io.ObjectOutputStream;
+import java.io.Serializable;
+import java.lang.reflect.Method;
+import java.util.concurrent.Callable;
+
+public class IsolatedClassloaderWorker implements Worker {
+    private final GroovySystemLoaderFactory groovySystemLoaderFactory = new GroovySystemLoaderFactory();
+    private final ClassLoaderStructure classLoaderStructure;
+    private final ClassLoader workerInfrastructureClassloader;
+    private final ServiceRegistry serviceRegistry;
+    private ClassLoader workerClassLoader;
+    private boolean reuseClassloader;
+
+    public IsolatedClassloaderWorker(ClassLoaderStructure classLoaderStructure, ClassLoader workerInfrastructureClassloader, ServiceRegistry serviceRegistry) {
+        this.classLoaderStructure = classLoaderStructure;
+        this.workerInfrastructureClassloader = workerInfrastructureClassloader;
+        this.serviceRegistry = serviceRegistry;
+    }
+
+    public IsolatedClassloaderWorker(ClassLoaderStructure classLoaderStructure, ClassLoader workerInfrastructureClassloader, ServiceRegistry serviceRegistry, boolean reuseClassloader) {
+        this(classLoaderStructure, workerInfrastructureClassloader, serviceRegistry);
+        this.reuseClassloader = reuseClassloader;
+    }
+
+    public DefaultWorkResult execute(ActionExecutionSpec spec) {
+        ClassLoader workerClassLoader = getWorkerClassLoader();
```

Comment:
*[draft]*

I suspect it's unnecessary to create another ClassLoader here, when running in the daemon process. Each layer adds a performance hit, and I think this ClassLoader currently ends up with an absolutely _enormous_ class path which makes class loading even slower.